### PR TITLE
Code update with working version

### PR DESCRIPTION
**Pull request by grandizzy**
- decoupled heap from loan definitions (renamed `LoansHeap` and its test as generic `Heap`, replaced `borrower` with `id` and `tp` with `val` )
- `insert` removed to `upsert` (create node if doesn't exist, update node if already in heap)
- keep track of nodes count in `count` variable (rather than relying on array's length) 
- `extractById` renamed to `remove`
- uncomment failing tests

### CODE REVIEW

File: src/libraries/Heap.sol [hunk 1]
```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.14;
+
+library Heap {
+
+    uint256 constant ROOT_INDEX = 1;
+
+    struct Data {
```
Comment (ith-harvey) [inline src/libraries/Heap.sol:8]: Why do we need the Data struct? Why not store it globally in the `Heap` class?

File: src/libraries/Heap.sol [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    function _bubbleDown(Data storage self_, Node memory n_, uint i_) private {
+        uint256 length = self_.count;
+        uint cIndex = i_ * 2; // left child index
+
+        if (length <= cIndex) {
+            _insert(self_, n_, i_);
+        } else {
+            Node memory largestChild = self_.nodes[cIndex];
```
Comment (ith-harvey) [inline src/libraries/Heap.sol:140]: It's not guaranteed that the left child is always the `largestChild`. There are scenarios where the right child is larger than the left.

I suggest renaming this.

File: src/libraries/Heap.sol [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+            if (length > cIndex+1 && self_.nodes[cIndex+1].val > largestChild.val) {
```
Comment (ith-harvey) [inline src/libraries/Heap.sol]: cleanup spacing here.

**Review by ith-harvey — COMMENTED** [review]
Is `totalBorrowers` in the Pool class the same value as `count` in the Heap?

Needs natspec in the code.

File: src/libraries/Heap.sol (hunk 1 above)
Comment (grandizzy) [reply src/libraries/Heap.sol:8]: that's because `Heap` is a library and cannot have storage

**Comment by grandizzy** [comment]
> Is `totalBorrowers` in the Pool class the same value as `count` in the Heap?
> 

The `totalBorrowers`  in the Pool is `count - 1`, good call for removing `totalBorrowers` entirely and keep track of borrowers from heap

> Needs natspec in the code.

**Review by MikeHathaway — APPROVED** [review]
LGTM

**Pull request merged**